core/retries/src/main/java/software/amazon/awssdk/retries/internal/circuitbreaker/TokenBucket.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.retries.internal.circuitbreaker;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.utils.ToString;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * A lock-free implementation of a token bucket. Tokens can be acquired from the bucket as long as there is sufficient capacity in
+ * the bucket.
+ */
+@SdkInternalApi
+public final class TokenBucket {
```

Comment:
There doesn't seem to be any test coverage for this class, which makes me nervous. Can we add some or repurpose the ones for the existing `TokenBucket`?

Reply:
I can add a specific test for this class but the coverage is already over 86% for instructions and 75% for branches as tested by the `StandardRetryStrategyTest` and `StandardRetryStrategyMiscTest` tests. Most of the missing coverage is from `toString`.